### PR TITLE
[SPARK-42891][CONNECT][PYTHON] Implement CoGrouped Map API

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Implement CoGrouped Map API: `applyInPandas`.

### Why are the changes needed?
Parity with vanilla PySpark.

### Does this PR introduce _any_ user-facing change?
Yes. CoGrouped Map API is supported as shown below.

```sh
>>> import pandas as pd
>>> df1 = spark.createDataFrame(
...   [(20000101, 1, 1.0), (20000101, 2, 2.0), (20000102, 1, 3.0), (20000102, 2, 4.0)], ("time", "id", "v1"))
>>> 
>>> df2 = spark.createDataFrame(
...   [(20000101, 1, "x"), (20000101, 2, "y")], ("time", "id", "v2"))
>>> 
>>> def asof_join(l, r):
...   return pd.merge_asof(l, r, on="time", by="id")
... 
>>> df1.groupby("id").cogroup(df2.groupby("id")).applyInPandas(
...   asof_join, schema="time int, id int, v1 double, v2 string"
... ).show()
+--------+---+---+---+
|    time| id| v1| v2|
+--------+---+---+---+
|20000101|  1|1.0|  x|
|20000102|  1|3.0|  x|
|20000101|  2|2.0|  y|
|20000102|  2|4.0|  y|
+--------+---+---+---+
```

### How was this patch tested?
Parity unit tests.

SPARK-41661

### CODE REVIEW

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 1]
```diff
@@ -509,6 +511,26 @@ class SparkConnectPlanner(val session: SparkSession) {
       .logicalPlan
   }
 
+  private def transformCoGroupMap(rel: proto.GroupMap): LogicalPlan = {
+    val pythonUdf = transformPythonUDF(rel.getFunc)
+
+    val input_cols =
```
Comment (LuciferYang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:517]: For Scala code, I suggest using Camel-Case

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 2]
```diff
@@ -509,6 +511,26 @@ class SparkConnectPlanner(val session: SparkSession) {
...
+      rel.getInputGroupingExpressionsList.asScala.toSeq.map(expr =>
+        Column(transformExpression(expr)))
+    val other_cols =
```
Comment (LuciferYang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:520]: ditto

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 3]
```diff
@@ -509,6 +511,26 @@ class SparkConnectPlanner(val session: SparkSession) {
       .logicalPlan
   }
 
+  private def transformCoGroupMap(rel: proto.GroupMap): LogicalPlan = {
```
Comment (LuciferYang) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:514]: `rel: proto.GroupMap` -> `rel: proto.CoGroupMap`

Comment (xinrong-meng) [reply connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:514]: Thanks.

**Comment by xinrong-meng** [comment]
May I get a review please @zhengruifeng @HyukjinKwon ?

**Comment by xinrong-meng** [comment]
cc [LuciferYang](https://github.com/LuciferYang) thanks!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.4.

**Comment by HyukjinKwon** [comment]
It has a conflict w/ branch-3.4. mind creating a backport PR please?

**Comment by LuciferYang** [comment]
late LGTM

File: python/pyspark/sql/connect/plan.py [hunk 4]
```diff
@@ -1950,6 +1950,46 @@ def plan(self, session: "SparkConnectClient") -> proto.Relation:
...
+    """Logical plan object for a CoGroup Map API: applyInPandas."""
+
+    def __init__(
+        self,
+        input: Optional["LogicalPlan"],
+        input_grouping_cols: Sequence[Column],
+        other: Optional["LogicalPlan"],
+        other_grouping_cols: Sequence[Column],
+        function: "UserDefinedFunction",
+        cols: List[Column],
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/plan.py:1963]: @xinrong-meng just realised that we don't use this variable. Should we remove? or should it be assigned to somewhere?

**Pull request closed without merging**